aten/src/ATen/OpaqueTensorImpl.h

```diff
@@ -77,35 +77,56 @@ struct CAFFE2_API OpaqueTensorImpl : public TensorImpl {
-//
-// NOTE: `allow_tensor_metadata_change` determines whether the TensorImpl shallow-copy
-// allows changes to its metadata (e.g. sizes / strides / storage / storage_offset).
-// See NOTE [ Metadata Change for a Detached Tensor ] for details.
-c10::intrusive_ptr<TensorImpl> shallow_copy_and_detach(
-    const c10::VariableVersion& version_counter,
-    bool allow_tensor_metadata_change) const override {
-  //AT_ASSERT(false);
-  auto impl = c10::make_intrusive<OpaqueTensorImpl<OpaqueHandle>>(
-    type_id(), dtype(), device(), opaque_handle_, sizes_);
-  // TensorImpl general fields
-  // Note that some of these fields are not used in opaque tensor code,
-  // and we copy them here only for completeness.
-  impl->sizes_ = sizes_;
-  impl->strides_ = strides_;
-  impl->storage_offset_ = storage_offset_;
-  impl->is_contiguous_ = is_contiguous_;
-  impl->is_wrapped_number_ = is_wrapped_number_;
-  impl->reserved_ = reserved_;
-  impl->set_version_counter(version_counter);
-  impl->set_allow_tensor_metadata_change(allow_tensor_metadata_change);
-
-  // OpaqueTensorImpl-specific fields (none currently).
-  return impl;
-}
+  /**
+   * Copy the storage pointer and the tensor metadata fields (e.g. sizes / strides / storage_offset)
+   * from one TensorImpl to another TensorImpl.
+   *
+   * For usage of `version_counter` and `allow_tensor_metadata_change`, see NOTE [ TensorImpl Shallow-Copying ].
+   */
+  void copy_tensor_data(
+      const TensorImpl* src_impl,
+      TensorImpl* dest_impl,
+      const c10::VariableVersion& version_counter,
+      bool allow_tensor_metadata_change) const override {
+    TensorImpl::copy_tensor_data(src_impl, dest_impl, version_counter, allow_tensor_metadata_change);
+
+    // OpaqueTensorImpl-specific fields.
+    auto src_opaque_impl = static_cast<const OpaqueTensorImpl<OpaqueHandle>*>(src_impl);
+    auto dest_opaque_impl = static_cast<OpaqueTensorImpl<OpaqueHandle>*>(dest_impl);
+    dest_opaque_impl->opaque_handle_ = src_opaque_impl->opaque_handle_;
+  }
+
+  /**
+   * Return a TensorImpl that is a shallow-copy of this TensorImpl.
+   *
+   * For usage of `version_counter` and `allow_tensor_metadata_change`,
+   * see NOTE [ TensorImpl Shallow-Copying ].
+   */
+  c10::intrusive_ptr<TensorImpl> shallow_copy_and_detach(
+      const c10::VariableVersion& version_counter,
+      bool allow_tensor_metadata_change) const override {
+    //AT_ASSERT(false);
+    auto impl = c10::make_intrusive<OpaqueTensorImpl<OpaqueHandle>>(
+      type_id(), dtype(), device(), opaque_handle_, sizes_);
+    copy_tensor_data(
+      /*src_impl=*/this,
+      /*dest_impl=*/impl.get(),
+      /*version_counter=*/version_counter,
+      /*allow_tensor_metadata_change=*/allow_tensor_metadata_change);
+    return impl;
+  }
+
+  /**
+   * Shallow-copies data from another TensorImpl into this TensorImpl.
+   */
+  void shallow_copy_from(c10::intrusive_ptr<TensorImpl> impl) override {
+    copy_tensor_data(
```

Comment:
this doesn't check allow_tensor_metadata_change, right?  Is that intentional?

Reply:
I added comment at https://github.com/pytorch/pytorch/pull/17072/files#diff-7f54d357d87885923a1536316f17ac6bR869 to explain this behavior.